packages/core/src/sdJwt/disclosures.ts

```diff
@@ -48,17 +51,65 @@ export class Disclosure {
             : [this.salt, this.value]
     }
 
-    public async digest(hasher: Hasher) {
-        // Memoize value so we don't have to re-compute
-        if (!this._digest) {
-            const hash = await hasher(this.encoded)
-            this._digest = Base64url.encode(hash)
+    public get digest() {
+        return this.#digest
+    }
+
+    /**
+     * Set the digest of the disclosure.
+     *
+     * NOTE: this method statically sets the digest, and does not verify whether the digest is correct.
+     * If you want to calculate the digest, use the `withCalculateDigest` method instead.
+     */
+    public withDigest(digest: string): DisclosureWithDigest {
+        this.#digest = digest
+        return this as DisclosureWithDigest
+    }
+
+    public withCalculateDigest<HasherImplementation extends Hasher>(
+        hasher: HasherImplementation,
+        // Whether to recalculate the digest, even if it is already set
+        { recalculate = false }: { recalculate?: boolean } = {}
+    ): WithCalculateDigestReturnType<HasherImplementation> {
+        // NOTE: the implementation of this method seems overly complex, but it allows
+        // us to return a promise if the hasher returns a promise, and a value otherwise.
+        // This allows this method to be used in environments where the calling scope
+        // is not async, as long as the hasher is not async either.
+        if (!recalculate && isDisclosureWithDigest(this)) {
+            return this as unknown as WithCalculateDigestReturnType<HasherImplementation>
         }
 
-        return this._digest
+        // Calculate digest
+        const hashResult = hasher(this.encoded)
+
+        // If promise, wait for it to resolve
+        if ('then' in hashResult) {
```

Comment:
I have a slight preference for `hashResult.constructor.name === "AsyncFunction"` to detect an async function. Not too big of a difference, so take it or leave it ;).

Reply:
Ah yes i was meaning to look up the proper way to do it. This was my first try and TS seemed to play well with it.

I'll update it 👍